core/base/mtx_io.cpp

```diff
@@ -760,6 +763,187 @@ matrix_data<ValueType, IndexType> read_raw(std::istream& is)
+    constexpr auto value_bit =
+        is_double ? 'D' : (is_float ? 'S' : (is_complex_double ? 'Z' : 'C'));
+    constexpr uint64 shift = 256;
+    constexpr uint64 type_bits = index_bit * shift + value_bit;
+    return 'G' +
+           shift *
+               ('I' +
+                shift *
+                    ('N' +
+                     shift *
+                         ('K' +
+                          shift * ('G' + shift * ('O' + shift * type_bits)))));
+}
+
+
+namespace {
+
+
+template <bool first>
+struct select_helper {};
+
+template <>
+struct select_helper<true> {
+    template <typename T1, typename T2>
+    static T1 get(T1 val, T2)
+    {
+        return val;
+    }
+};
+
+template <>
+struct select_helper<false> {
+    template <typename T1, typename T2>
+    static T2 get(T1, T2 val)
+    {
+        return val;
+    }
+};
+
+
+template <typename FileValueType, typename FileIndexType, typename ValueType,
+          typename IndexType>
+matrix_data<ValueType, IndexType> read_binary_convert(std::istream& is,
+                                                      uint64 num_rows,
+                                                      uint64 num_cols,
+                                                      uint64 num_entries)
+{
+    if (num_rows > std::numeric_limits<IndexType>::max() ||
+        num_cols > std::numeric_limits<IndexType>::max()) {
+        throw GKO_STREAM_ERROR(
+            "cannot read into this format, its index type would overflow");
+    }
+    if (is_complex<FileValueType>() && !is_complex<ValueType>()) {
+        throw GKO_STREAM_ERROR(
+            "cannot read into this format, would assign complex to real");
+    }
+    matrix_data<ValueType, IndexType> result(gko::dim<2>{num_rows, num_cols});
+    result.nonzeros.resize(num_entries);
+    constexpr auto entry_binary_size =
+        sizeof(FileValueType) + 2 * sizeof(FileIndexType);
+    for (size_type i = 0; i < num_entries; i++) {
+        std::array<char, entry_binary_size> block;
+        GKO_CHECK_STREAM(is.read(block.data(), entry_binary_size),
+                         "failed reading entry " + std::to_string(i));
+        FileValueType value{};
+        FileIndexType row{};
+        FileIndexType column{};
+        std::memcpy(&row, &block[0], sizeof(FileIndexType));
+        std::memcpy(&column, &block[sizeof(FileIndexType)],
+                    sizeof(FileIndexType));
+        std::memcpy(&value, &block[2 * sizeof(FileIndexType)],
+                    sizeof(FileValueType));
+        result.nonzeros[i].value = static_cast<ValueType>(
+            select_helper<is_complex<ValueType>()>::get(value, real(value)));
+        result.nonzeros[i].row = row;
+        result.nonzeros[i].column = column;
+    }
+    return result;
+}
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+matrix_data<ValueType, IndexType> read_binary_raw(std::istream& is)
+{
+    std::array<char, 32> header{};
```

Comment:
Are we sure the header will always fit into 32 characters ?

Reply:
yes, because we hand-fill it via memcpy, it will always fit into sizeof(uint64_t) * 4 chars. But I'll add a static_assert just to make sure :)